### PR TITLE
Serial Terminal

### PR DESCRIPTION
**Pull request by rwalton-arm**
### Description
Implement serial terminal and sterm command.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: tests/sterm/test_terminal.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+
+@pytest.mark.parametrize("menu_key", INVALID_MENU_KEYS)
+def test_ignores_invalid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key)
+
+    with pytest.raises(StopIteration):
```
Comment (Patater) [inline tests/sterm/test_terminal.py:142]: Why is a StopIteration an invalid menu key? Why not KeyboardError or something?

File: tests/sterm/test_terminal.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+    term.handle_menu_key.assert_not_called()
+
+
+def test_reset_sends_serial_break(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    term.reset()
+
+    mock_serial().sendBreak.assert_called_once()
```
Comment (Patater) [inline tests/sterm/test_terminal.py:153]: Do we really need to send break conditions? Is this a reliable way to reset a target? @jamesbeyond Do you know if we have any targets that reset when given a serial break? Is this a common thing?

File: tests/sterm/test_terminal.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def test_ctrl_b_sends_reset_to_serial_port(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_B, terminal.CTRL_C)
```
Comment (Patater) [inline tests/sterm/test_terminal.py:159]: Would this test pass if we did nothing for CTRL_B and made CTRL_C send a break to the serial port? How do we know it's the CTRL_B that's giving us the desired effect and not something else? I am curious why we sent CTRL_C. I guess it is to cause the console to stop blocking, but perhaps there is a more targeted way to do that.

File: tests/sterm/test_terminal.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+def test_starts_sterm_thread(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().start.assert_called_once()
+
+
+def test_joins_tx_and_rx_threads(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_any_call(True)
```
[the remark this answers is not in the document]
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:59]: `join(True)` means "join the tx and rx threads" - `Miniterm` handles the thread management after that so it seemed good enough to just check we called `join` correctly here.

File: tests/sterm/test_terminal.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def test_joins_tx_thread_after_keyboard_interrupt(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (KeyboardInterrupt(), None)
+
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_called_with()
```
[the remark this answers is not in the document]
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:67]: No they aren't equivalent,
 `assert_any_call` means "if the mock was ever called with `args`"
 `assert_called_with` means "if the most recent call was with `args`"

[the remark this answers is not in the document]
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:67]: We check sterm closes after any exception below.

File: tests/sterm/test_terminal.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+def test_joins_tx_and_rx_threads(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_any_call(True)
+
+
+def test_joins_tx_thread_after_keyboard_interrupt(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (KeyboardInterrupt(), None)
+
+    terminal.run("tty.122", 9600)
```
[the remark this answers is not in the document]
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:65]: `KeyboardInterrupt` is handled by `terminal.run` so it doesn't bubble up and fail the test.

File: tests/sterm/test_terminal.py (hunk 2 above)
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:153]: Not sure, this was just what the old tool did. Serial break does reset my DISCO_L475 though.

File: tests/sterm/test_terminal.py (hunk 3 above)
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:159]: CTRL_C doesn't send a break to the serial port, it tells the terminal to stop running and blocking. Maybe it's better to just handle the `StopIteration` like I did in the above test, to make it clearer we're not testing CTRL_C handling. I did think the `StopIteration` handling was a bit dirty though...

File: tests/sterm/test_terminal.py (hunk 1 above)
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:142]: `StopIteration` isn't a valid menu key, it's a way to get the terminal to stop blocking so I can assert on the state of the mock.

File: tests/sterm/test_terminal.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:67]: Actually, below we only check `Exception` derived exceptions. I'll add a test for close on `KeyboardInterrupt`.

File: tests/sterm/test_terminal.py (hunk 2 above)
Comment (rwalton-arm) [reply tests/sterm/test_terminal.py:153]: Note: If we didn't send serial break and change the menu text and key handling, we wouldn't need to subclass `Miniterm` at all. However, we will change the UX if we do that.

Comment (Patater) [reply tests/sterm/test_terminal.py:153]: Alright. Learn a new feature everyday. If we run --sterm without --flash, I can see it being useful that the target would be reset. I would have expected a separate reset mechanism than a serial break condition, but it works how it works.

**Review by Patater — APPROVED** [review]

**Review by jainvikas8 — APPROVED** [review]

**Pull request merged**